Add Class and Class Element Decorators and `accessor` Keyword

pzuraq opened the pull request:
This PR includes the spec changes for the current iteration of the [decorators proposal](https://github.com/tc39/proposal-decorators). This proposal is still in progress, currently in stage 2, and the PR is mainly being opened for reviewing and iteration purposes.

pzuraq committed:
Add Class and Class Element Decorators and `accessor` Keyword

Updates

- Refactor to use ClassElementDefinition more extensively
- Separate parsing class elements and defining them so that we can
  control the order of definition and decorator application
- Extract and centralize class element decoration into a single place
- Fix completions in ClassDefinitionEvaluation (need to reset the env
  after each possible abrupt completion).
- Refactor adding private methods to instance to pull directly from
  `Class.[[Elements]]`, so we don't have multiple sources of truth for
  the elements that are defined on a class. `Class.[[Elements]]` is the
  canonical source of truth after ClassDefinitionEvaluation, and any
  operations afterward such as instantiation should base themselves on
  that.
- Refactor to use anonymous function syntax.

Non-normative fixes

Fixes based on feedback from @jmdyck

Remove dynamic [[HomeObject]] from decorator return values

Throw an error if the value passed to `addInitializer` is not callable

Set the name of the `addInitializer` function

Remove setFunctionName from decorator application

Call decorators with their natural `this` value.

For more details, see this issue: https://github.com/tc39/proposal-decorators/issues/487

Reverse initializer order

Update field and accessor extra initializers to run after definition

tomdale commented on spec.html:
```diff
@@ -5784,20 +5955,25 @@ <h1>PrivateSet ( _P_, _O_, _value_ )</h1>
-          1. Assert: ! IsPropertyKey(_fieldName_) is *true*.
-          1. Perform ? CreateDataPropertyOrThrow(_receiver_, _fieldName_, _initValue_).
+        1. Let _initializers_ be _elementRecord_.[[Initializers]].
+        1. If _initializers is not an empty list, then
```
Missing closing `_` after `initializers`

leobalter commented on spec.html:
```diff
@@ -7884,9 +8060,12 @@ <h1>Runtime Semantics: NamedEvaluation</h1>
+      <emu-grammar>ClassExpression : DecoratorList? `class` ClassTail</emu-grammar>
       <emu-alg>
-        1. Let _value_ be the result of ClassDefinitionEvaluation of |ClassTail| with arguments *undefined* and _name_.
+        1. If |DecoratorList_opt| is present, then
```
```suggestion
        1. If |DecoratorList| is present, then
```

here and in other similar places with `|DecoratorList_opt| is present`.

Reference for consistency: https://tc39.es/ecma262/#sec-static-semantics-lexicallydeclarednames

leobalter replied:
There are [some](https://tc39.es/ecma262/#sec-runtime-semantics-propertydestructuringassignmentevaluation) [few](https://tc39.es/ecma262/#sec-runtime-semantics-classfielddefinitionevaluation) [parts](https://tc39.es/ecma262/#sec-runtime-semantics-classdefinitionevaluation) where `_opt` is preserved, but it feels like an error.

jmdyck commented:
Rather than make a ton of "suggested changes" here, I've made a [PR]( https://github.com/pzuraq/ecma262/pull/2) against the pzuraq:decorators branch.

jmdyck commented on spec.html:
```diff
@@ -21534,47 +22137,64 @@ <h1>Runtime Semantics: ClassDefinitionEvaluation</h1>
+          1. If IsStatic of _e_ is *false*, append _element_ to _instanceElements_.
+          1. Else, append _element_ to _staticElements_.
+        1. Let _parentInstanceMetadata_ be ? Get(_protoParent_, @@metadata).
+        1. Perform ! DefinePropertyOrThrow(_proto_, @@metadata, ConvertDecoratorMetadataListToObject(_instanceMetadataList_, _parentInstanceMetadata_)).
```
ConvertDecoratorMetadataListToObject returns an Object, but DefinePropertyOrThrow expects a Property Descriptor there.

pzuraq replied:
Updated to use `CreateDataPropertyOrThrow`

jmdyck commented on spec.html:
```diff
@@ -21534,47 +22137,64 @@ <h1>Runtime Semantics: ClassDefinitionEvaluation</h1>
             1. Set the running execution context's PrivateEnvironment to _outerPrivateEnvironment_.
             1. Return _result_.
+        1. Let _parentStaticMetadata_ be ? Get(_constructorParent_, @@metadata).
+        1. Perform ! DefinePropertyOrThrow(_F_, @@metadata, ConvertDecoratorMetadataListToObject(_staticMetadataList_, _parentStaticMetadata_)).
```
(Ditto previous comment.)

jmdyck commented:
Since Decorators are only used with Class definitions, would it make sense to put the Decorators clause within the Class Definitions clause?

ljharb commented:
They’re only used with classes *for now*- there’s lots of interest in follow up proposals to add them in other places.

jmdyck commented:
(I did look for follow-up proposals, but only found a couple at stage 0.)

jmdyck commented on spec.html:
```diff
@@ -21416,28 +21894,131 @@ <h1>Static Semantics: ContainsArguments</h1>
+      <emu-alg>
+        1. Let _name_ be the result of evaluating |ClassElementName|.
+        1. ReturnIfAbrupt(_name_).
+        1. Let _privateStateDesc_ be the string-concatenation of *"storage for "* and the StringValue of |ClassElementName|.
```
StringValue isn't defined on ClassElementName, and I suspect it shouldn't be. (Currently, StringValue is only defined on Identifiers and similar; ClassElementName is a much wider syntax.)

jmdyck commented on spec.html:
```diff
@@ -21622,24 +22248,30 @@ <h1>Runtime Semantics: BindingClassDeclarationEvaluation</h1>
         1. Return NormalCompletion(~empty~).
       </emu-alg>
       <emu-note>
         <p><emu-grammar>ClassDeclaration : `class` ClassTail</emu-grammar> only occurs as part of an |ExportDeclaration| and is never directly evaluated.</p>
```
Generalize this to the new syntax? (Ditto the similar note 12 lines up.)

jmdyck reviewed (changes requested):
I believe the following SDOs need to be modified to handle the new syntax for ClassDeclaration, ClassExpression, and/or ClassElement:
- BoundNames
- ClassElementKind
- ComputedPropertyContains
- Early Errors
- IsConstantDeclaration
- IsFunctionDefinition
- IsStatic
- PrivateBoundIdentifiers
- PropName

(Note that currently, 4 of the 5 RHSs of ClassElement are chain productions, and so often have implicit definitions for SDOs, but they aren't chain productions in this PR, and so will need explicit definitions.)

senocular commented on spec.html:
```diff
@@ -21534,47 +22123,64 @@ <h1>Runtime Semantics: ClassDefinitionEvaluation</h1>
+        1. Perform ! DefinePropertyOrThrow(_proto_, @@metadata, ConvertDecoratorMetadataListToObject(_instanceMetadataList_, _parentInstanceMetadata_)).
         1. Set the running execution context's LexicalEnvironment to _env_.
         1. If _classBinding_ is not *undefined*, then
           1. Perform _classScope_.InitializeBinding(_classBinding_, _F_).
```
This should be set after class decorators run ([below](https://github.com/tc39/ecma262/pull/2417/files#diff-181371b08d71216599b0acccbaabd03c306da6de142ea6275c2135810999805aR22169-R22173)) to ensure this binding uses any replacement class returned by class decorators.

pzuraq replied:
This should be addressed now!

pzuraq commented on spec.html:
```diff
@@ -11418,16 +11597,25 @@ <h1>MakeMethod ( _F_, _homeObject_ )</h1>
+          1. Assert: _metadataList_ and _isStatic_ are present.
+          1. For each element _decorator_ of _decorators_, do
+            1. Let _context_ be CreateDecoratorContextObject(*"method"*, _key_, _metadataList_, _initializers_, _isStatic_).
+            1. Let _newMethod_ be ? Call(_decorator_.[[Decorator]], *undefined*, &laquo; _closure_, _context_ &raquo;).
```
No longer necessary with the removal of `@init:`

pzuraq commented on spec.html:
```diff
@@ -21026,16 +21232,287 @@ <h1>Runtime Semantics: Evaluation</h1>
+        1. Let _metadataList_ be _f_.[[MetadataList]].
+        1. Let _propertyKey_ be _f_.[[PropertyKey]].
+        1. Let _metadataForKey_ be the element of _metadataList_ whose [[Key]] is the same as _metadataKey_.
+        1. If _metadataForKey_ does not exist, then
```
A similar idiom exists in `ParseModule` and elsewhere. I originally asked about adding a Map construct to complement the List construct, since that is what is happening here, but after discussion with the editors it seemed like they preferred continuing to build on the List construct in this way.

jmdyck replied:
What do you think is a similar idiom in `ParseModule`?

pzuraq replied:
```
1. Let _ie_ be the element of _importEntries_ whose [[LocalName]] is the same as _ee_.[[LocalName]].
1. If _ie_.[[ImportName]] is *"\*"*, then
  1. NOTE: This is a re-export of an imported module namespace object.
  1. Append _ee_ to _localExportEntries_.
1. Else,
  1. NOTE: This is a re-export of a single name.
  1. Append the ExportEntry Record { [[ModuleRequest]]: _ie_.[[ModuleRequest]], [[ImportName]]: _ie_.[[ImportName]], [[LocalName]]: *null*, [[ExportName]]: _ee_.[[ExportName]] } to _indirectExportEntries_.
```

jmdyck replied:
I'm fine with `the element of <List> whose <field_name> is the same as <expr>`. But in `ParseModule`, such an element *exists*, so `_ie_` is well-defined. In your algorithm, there might not be such an element, so `_metadataForKey_` might be an alias for a lack of a result, which you check via `If _metadataForKey_ does not exist`. I'm saying that nowhere in the spec do we have `If <alias> does not exist` or `If <alias> exists`.

pzuraq commented on spec.html:
```diff
@@ -21446,24 +22027,46 @@ <h1>Runtime Semantics: ClassFieldDefinitionEvaluation</h1>
+      <emu-alg>
+        1. If |DecoratorList_opt| is present, then
+          1. Let _decorators_ be ? DecoratorListEvaluation of |DecoratorList|.
+        1. Else, let _decorators_ be a new empty List.
```
I've tried to do this by creating a `MaybeDecoratorList` grammar, let me know if this solves the issue

jmdyck replied:
(I got a notification for the above comment, but I didn't remember the context after 7 months, so I clicked the link, but GitHub wouldn't show me this conversation because you'd already marked it resolved. I had to click a bunch of "Show resolved" before I found it. For this reason, and others, I think it's better to allow the *reviewer* to mark a conversation as resolved. I believe that's the local convention, though I couldn't find anything to that effect in tc39/how-we-work.)

pzuraq replied:
Sorry, part of my workflow in large PRs is usually to use these comments as a todo list, and to mark as resolved when I've checked this review off the list. I'll unresolve these.

As for this issue, `MaybeDecoratorList` didn't actually work because it'll fail to parse in a lot of situations, `DecoratorList` needs to be truly optional. I can't find an example of an SDO where the left hand side of the SDO itself is optional, and if the whole production doesn't exist it returns a different value, could use some advice on the best way to do this.

jmdyck replied:
I'll think about it more in my next review (if it hasn't already been resolved by then).

pzuraq commented on spec.html:
```diff
@@ -4173,13 +4190,167 @@ <h1>The ClassFieldDefinition Record Specification Type</h1>
+
+    <emu-clause id="sec-metadata-record-specification-type">
+      <h1>The Metadata Record Specification Type</h1>
+      <p>The Metadata type is a Record used in the specification of decorator metadata. When defining metadata for a given decorated value, one Metadata Record is created for each user unique provided metadata key.</p>
```
Wording updated, agreed it was awkward in that order

pzuraq commented on spec.html:
```diff
@@ -21026,16 +21232,287 @@ <h1>Runtime Semantics: Evaluation</h1>
+            1. Set _publicMetadata_ to the Record { [[Key]]: _propertyKey_, [[Value]]: _value_ }.
+            1. Append _publicMetadata_ to _metadataForKey_.[[Public]].
+          1. Else, set _publicMetadata_.[[Value]] to _value_.
+        1. If _kind_ is ~private~, then
```
Addressed, I agree these are mutually exclusive branches so the change makes sense

pzuraq commented on spec.html:
```diff
@@ -21026,16 +21232,287 @@ <h1>Runtime Semantics: Evaluation</h1>
+            1. Let _length_ be the number of non-optional parameters of a GetDecoratorMetadata function as specified below.
+            1. Let _closure_ be ! CreateBuiltinFunction(_steps_, _length_, *""*, &laquo; [[OwnPrivateMetadata]] &raquo;).
+            1. Set _closure_.[[OwnPrivateMetadata]] to _privateMetadata_.
+            1. Let _desc_ be the PropertyDescriptor { [[Get]]: _closure_, [[Enumerable]]: _enumerable_, [[Configurable]]: *true* }.
```
Fixed

pzuraq commented on spec.html:
```diff
@@ -21026,16 +21232,287 @@ <h1>Runtime Semantics: Evaluation</h1>
+    </emu-clause>
+
+    <emu-clause id="sec-createdecoratorcontextobject" aoid="CreateDecoratorContextObject">
+      <h1>CreateDecoratorContextObject ( _kind_, _name_, _metadataList_, _initializers_ [ , _isStatic_ ] )</h1>
```
Forgot to add that, updated with the latest version of the spec, it now adds `addInitializer` to the context object.

pzuraq commented on spec.html:
```diff
@@ -21416,28 +21894,131 @@ <h1>Static Semantics: ContainsArguments</h1>
+        1. Let _scope_ be the LexicalEnvironment of the running execution context.
+        1. Let _privateScope_ be the running execution context's PrivateEnvironment.
+        1. Let _sourceText_ be the empty sequence of Unicode code points.
+        1. Let _initializer_ be ! OrdinaryFunctionCreate(%Function.prototype%, _sourceText_, _formalParameterList_, _Initializer_, ~non-lexical-this~, _scope_, _privateScope_).
```
There are other places where we do this currently in the spec, see `ClassFieldDefinitionEvaluation`, so I was following that idiom

jmdyck replied:
In `ClassFieldDefinitionEvaluation`, the capitalized `Initializer` isn't an alias, it's a nonterminal, so it has to be capitalized. Here, the capitalized `_Initializer_` is an alias, so should not be capitalized. (The spec does have capitalized aliases, but most are only one or two letters long -- the exceptions are uncommon and I don't think we should add to them.)

Re the more general problem: Throughout the spec, an "Initializer" is a Parse Node (an instance of the *Initializer* nonterminal). But in the context of class fields, it also refers to a function object that, when called, returns an initial value for a class field. I think these two uses should be distinguished, by changing the latter to "InitializingFunction" or "InitializerFunction".

pzuraq commented on spec.html:
```diff
@@ -21026,16 +21232,287 @@ <h1>Runtime Semantics: Evaluation</h1>
+    </emu-clause>
+
+    <emu-clause id="sec-createdecoratoraccessobject" aoid="CreateDecoratorAccessObject">
+      <h1>CreateDecoratorAccessObject ( _kind_, _name_ )</h1>
```
Fixed, that call should have used the local `_kind_`.

pzuraq commented:
> I believe the following SDOs need to be modified to handle the new syntax for ClassDeclaration, ClassExpression, and/or ClassElement:
> 
> * BoundNames
> * ClassElementKind
> * ComputedPropertyContains
> * Early Errors
> * IsConstantDeclaration
> * IsFunctionDefinition
> * IsStatic
> * PrivateBoundIdentifiers
> * PropName
> 
> (Note that currently, 4 of the 5 RHSs of ClassElement are chain productions, and so often have implicit definitions for SDOs, but they aren't chain productions in this PR, and so will need explicit definitions.)

@jmdyck regarding this comment, I'm not quite sure I understand what needs to be done here. Why do these productions need to be updated exactly?

jmdyck commented:
Consider the production `ClassElement : FieldDefinition` in the status quo. It's a chain production, so (e.g.) the SDO ComputedPropertyContains doesn't need to have a definition for that production; instead we just chain down to FieldDefinition, for which ComputedPropertyContains does have an explicit definition.

But this PR replaces that production with `ClassElement : DecoratorList? FieldDefinition`, which is not a chain production, so there needs to be an explicit definition of ComputedPropertyContains for that production.

And likewise for the other alternatives of `ClassElement` that now have `DecoratorList?`, and likewise for the other SDOs that reach ClassElement.

Jack-Works commented on spec.html:
```diff
@@ -24637,17 +25241,50 @@ <h1>
+          1. Let _initializer_ be ! CreateFieldInitializerFunction(_homeObject_, _name_, |Initializer|).
+          1. Append _initializer_ to _initializers_.
+        1. For each element _decorator_ of _decorators_, do
+          1. Let _context_ be CreateDecoratorContextObject(*"accessor"*, _name_, _metadataList_, _extraInitializers_, _isStatic_).
```
what's the difference between `extraInitializers` and `initialiers`? Why some functions is appended to the prior one and some are appended to the latter one?

pzuraq replied:
Accessors have two initialization steps:

1. The per-instance extraInitializers are run when the class instance is created, _before_ class fields are assigned. This is the same time that method initializers in general are run.
2. The private field backing the accessor is initialized. This runs during class field initialization.

Accessors are essentially sugar for declaring a getter/setter + field, so these two steps can be thought of as the combination of the initialization for both of those types of values, e.g.:

```js
class Foo {
  @dec
  #bar = 123;

  @dec
  get bar() {
    return this.#bar;
  }
}
```

In this example, `@dec` would be able to add an initializer for the `#bar` private field which would run when the field is defined, and separately add an initializer for the `get bar()` getter which would run at the beginning of class instantiation. Decorating an `accessor` lets you do both.

Jack-Works commented on spec.html:
```diff
@@ -24694,27 +25331,50 @@ <h1>
       <h1>
         Runtime Semantics: ClassElementEvaluation (
           _object_: unknown,
+          _metadataList: a List of Metadata Records,
```
```suggestion
          _metadataList_: a List of Metadata Records,
```

Jack-Works commented on spec.html:
```diff
@@ -24166,16 +24373,319 @@ <h1>Runtime Semantics: Evaluation</h1>
+        1. Let _f_ be the active function object.
+        1. Return ? PrivateGet(_f_.[[PrivateStateKey]], _o_).
+      </emu-alg>
+      <p>A PrivateValueSetter function is an anonymous built-in function with a [[PrivateStateKey]] internal slot. When a PrivateValueSetter function is called with argument _value_ it performs the following steps:</p>
```
Can you use Abstract Closure to rewrite those built-in functions if possible? It will reduce the need for internal slots that are used to capture variables. An example: https://tc39.es/ecma262/multipage/control-abstraction-objects.html#sec-asyncgeneratorawaitreturn

```
1. Let _privateValueGetter_ be a new Abstract Closure with parameter (_value_) that captures _name_ and perform the following steps when called:
    1. Let _o_ be the *this* value.
    2. Perform ? PrivateSet(_name_, _o_, _value_).
    3. Return *undefined*
1. Let _f_ be ! CreateBuiltinFunction(_privateValueGetter_, 1, "", <<>>).
```

pzuraq commented on spec.html:
```diff
@@ -24166,16 +24373,319 @@ <h1>Runtime Semantics: Evaluation</h1>
+        `@` DecoratorMemberExpression[?Yield, ?Await]
+        `@` DecoratorCallExpression[?Yield, ?Await]
+
+      DecoratorMemberExpression[Yield, Await] :
```
No, the intention is to support only simple chaining followed by a call at the end, OR any expression inside parens:

```js
// valid
@id
@id.id
@id()
@id.id()
@(id['foo'].bar().baz())

// invalid
@id().id
@id().id()
@(id['foo'])()
```

Jack-Works replied:
As the current syntax, I can write `@(getDecorator('f')).a`. Is that right or I read the syntax wrong?

pzuraq replied:
That is not the intention, but it's definitely possible, I copied the decorator syntax from an earlier version of the proposal (the one proposed in 2019) which was supposed to have these constraints, will try to think though how to update the syntax

pzuraq replied:
I believe this should be fixed

keithamus commented on spec.html:
```diff
@@ -24623,10 +25134,103 @@ <h1>Static Semantics: ContainsArguments</h1>
+        1. Perform SetFunctionName(_getter_, _name_, *"get"*).
+        1. Return _getter_.
+      </emu-alg>
+      <p>An AutoAccessorGetter function is an anonymous built-in function with a [[PrivateStateKey]] internal slot. When an AutoAccessorGetter function that expects no arguments is called it performs the following steps:</p>
```
What happens when an AutoAccessorGetter _does_ expect arguments.

ljharb replied:
i think this will get cleaned up when it's converted to use abstract closures.

keithamus commented on spec.html:
```diff
@@ -24191,15 +24701,16 @@ <h2>Syntax</h2>
 
       FieldDefinition[Yield, Await] :
         ClassElementName[?Yield, ?Await] Initializer[+In, ?Yield, ?Await]?
+        `accessor` ClassElementName[?Yield, ?Await] Initializer[+In, ?Yield, ?Await]?
```
Perhaps there's an issue covering this, but `accessor` is already a valid class element name so the following is valid grammar today:

```js
class Foo {
  accessor
  x = 1
}
```

Does this need `no [LineTerminator] here`?

pzuraq commented on spec.html:
```diff
@@ -24191,15 +24701,16 @@ <h2>Syntax</h2>
         `;`
 
       FieldDefinition[Yield, Await] :
         ClassElementName[?Yield, ?Await] Initializer[+In, ?Yield, ?Await]?
+        `accessor` ClassElementName[?Yield, ?Await] Initializer[+In, ?Yield, ?Await]?
```
I was following the same pattern as `\`static\`` here, it seems easier to handle it this way in the SDOs as well

pzuraq replied:
So I tried to do this and got a lot of errors saying that various places could not find the grammar. Same with combining `static` into an optional for methods and fields.

pzuraq replied:
I'm not sure if this is still relevant, if it is feel free to unresolve it

pzuraq commented:
Still working on addressing the feedback items, but I made a number of updates so far and wanted to get those updates up so folks could start reviewing sooner rather than later. Updates since last draft:

- Separate parsing class elements and defining them so that we can control the order of definition and decorator application. This also allows us to avoid large numbers of optional arguments for various SDOs, instead the SDOs are now responsible for returning a `ClassElementDefinition` record, and the invoker of the SDO is then responsible for actually decorating and defining the element.
- Run field decorators after method decorators, per https://github.com/tc39/proposal-decorators/issues/424#issuecomment-1032828707
- Extract and centralize class element decoration into a single abstract operation, so we don't have to worry about keeping it in sync in several different places.
- Fix completions in ClassDefinitionEvaluation (need to reset the env after each possible abrupt completion).
- Refactor adding private methods to instance to pull directly from `Class.[[Elements]]`, so we don't have multiple sources of truth for the elements that are defined on a class. `Class.[[Elements]]` is the canonical source of truth after ClassDefinitionEvaluation, and any operations afterward such as instantiation should base themselves on that.

  NOTE: This does mean that the abstract operation for instantiating private methods on the instance has to "run" and filter out the private methods from all of the other elements per-instantiation _in the spec_, but my assumption here was that the exact details of the spec wouldn't need to be followed and implementations could, for instance, filter out and store the private method definitions at class definition time and then copy them onto the instance all at once. If we need to make a note to that effect, we can, and if it's important for the spec to run the AO only at definition time for some reason we can do that, it's just a bit more complicated.
- Refactor to use anonymous function syntax.

Sidenote: It would be really nice to put comments inline in the spec to clarify the thinking for a specific portion, unsure if that's a feature anyone would be interested in?

ljharb commented on spec.html:
```diff
@@ -4671,7 +4685,10 @@ <h1>The ClassFieldDefinition Record Specification Type</h1>
+              [[Key]]
+            </td>
+            <td>
+              All
```
maybe these should be listed explicitly?

ljharb commented on spec.html:
```diff
@@ -6933,12 +7120,13 @@ <h1>
-        1. For each element _fieldRecord_ of _fields_, do
-          1. Perform ? DefineField(_O_, _fieldRecord_).
+        1. Let _elements_ be the value of _constructor_.[[Elements]].
+        1. Perform ? InitializePrivateMethodsAndAccessors
```
presumably this takes arguments?

ljharb commented on spec.html:
```diff
@@ -24179,16 +24361,356 @@ <h1>Runtime Semantics: Evaluation</h1>
+          1. Let _setterClosure_ be a new Abstract Closure with parameter (_value_) that captures _name_ and performs the following steps when called:
+            1. Let _o_ be the *this* value.
+            1. Perform ? PrivateSet(_name_, _o_, _value_).
+            1. Return *undefined*
```
```suggestion
            1. Return *undefined*.
```

ljharb commented on spec.html:
```diff
@@ -24179,16 +24361,356 @@ <h1>Runtime Semantics: Evaluation</h1>
+            1. Let _o_ be the *this* value.
+            1. Perform ? PrivateSet(_name_, _o_, _value_).
+            1. Return *undefined*
+          1. Let _setter_ be ! CreateBuiltinFunction(_setterClosure_, 1, "", &laquo; &raquo;).
```
```suggestion
          1. Let _setter_ be ! CreateBuiltinFunction(_setterClosure_, 1, *""*, &laquo; &raquo;).
```

ljharb commented on spec.html:
```diff
@@ -24179,16 +24361,356 @@ <h1>Runtime Semantics: Evaluation</h1>
+        1. Let _addInitializerClosure_ be a new Abstract Closure with parameter (_initializer_) that captures _initializers_ and performs the following steps when called:
+          1. Append _initializer_ to _initializers_.
+          1. Return *undefined*.
+        1. Let _addInitializer_ be ! CreateBuiltinFunction(_addInitializerClosure_, 1, "", &laquo; &raquo;).
+        1. Return _addInitializer_.
```
```suggestion
        1. Return ! CreateBuiltinFunction(_addInitializerClosure_, 1, *""*, &laquo; &raquo;).
```

ljharb commented on spec.html:
```diff
@@ -24179,16 +24361,356 @@ <h1>Runtime Semantics: Evaluation</h1>
+      </dl>
+      <emu-alg>
+        1. Let _setMetadataClosure_ be a new Abstract Closure with parameter (_metadataKey_, _value_) that captures _kind_, _metadataList_, and _propertyKey_ and performs the following steps when called:
+          1. If _metadataKey_ is not a Symbol, throw a *TypeError* exception.
```
```suggestion
          1. If Type(_metadataKey_) is not Symbol, throw a *TypeError* exception.
```

ljharb commented on spec.html:
```diff
@@ -24179,16 +24361,356 @@ <h1>Runtime Semantics: Evaluation</h1>
+              1. Append _privateMetadata_ to _metadataForKey_.[[Private]].
+            1. Else, set _privateMetadata_.[[Value]] to _value_.
+          1. Return *undefined*.
+        1. Let _setMetadata_ be ! CreateBuiltinFunction(_setMetadataClosure_, 2, *""*, &laquo; &raquo;).
+        1. Return _setMetadata_.
```
```suggestion
        1. Return CreateBuiltinFunction(_setMetadataClosure_, 2, *""*, &laquo; &raquo;).
```
(all  CreateBuiltinFunction invocations no longer return a completion record, and thus should not have a ! prefix)

ljharb commented on spec.html:
```diff
@@ -24179,16 +24361,356 @@ <h1>Runtime Semantics: Evaluation</h1>
+            1. Let _privateMetadata_ be the element in _metadataForKey_.[[Private]] whose [[Key]] is the same as _propertyKey_.
+            1. If _privateMetadata_ does not exist, return *undefined*.
+            1. Return _privateMetadata_.[[Value]].
+        1. Let _getMetadata_ be ! CreateBuiltinFunction(_getMetadataClosure_, _length_, *""*, &laquo; &raquo;).
+        1. Return _getMetadata_.
```
```suggestion
        1. Return CreateBuiltinFunction(_getMetadataClosure_, _length_, *""*, &laquo; &raquo;).
```

ljharb commented on spec.html:
```diff
@@ -24179,16 +24361,356 @@ <h1>Runtime Semantics: Evaluation</h1>
+        1. Else if _kind_ is ~getter~, let _kindStr_ be *"getter"*.
+        1. Else if _kind_ is ~setter~, let _kindStr_ be *"setter"*.
+        1. Else if _kind_ is ~accessor~, let _kindStr_ be *"accessor"*.
+        1. Else, let _kindStr_ be *"field"*.
```
```suggestion
        1. Else,
          1. Assert: _kind_ is ~field~.
          1. Let _kindStr_ be *"field"*.
```

ljharb commented on spec.html:
```diff
@@ -24636,6 +25156,102 @@ <h1>Static Semantics: ContainsArguments ( ): a Boolean</h1>
+        1. Let _getterClosure_ be a new Abstract Closure with no parameters that captures _privateStateName_ and perform the following steps when called:
+          1. Let _o_ be the *this* value.
+          1. Return ? PrivateGet(_privateStateName_, _o_).
+        1. Let _getter_ be ! CreateBuiltinFunction(_getterClosure_, 0, *""*, &laquo; &raquo;).
+        1. Perform MakeMethod(_getter_, _homeObject_).
+        1. Perform SetFunctionName(_getter_, _name_, *"get"*).
```
```suggestion
        1. Let _getter_ be ! CreateBuiltinFunction(_getterClosure_, 0, *"get"*, &laquo; &raquo;).
        1. Perform MakeMethod(_getter_, _homeObject_).
```

ljharb commented on spec.html:
```diff
@@ -24636,6 +25156,102 @@ <h1>Static Semantics: ContainsArguments ( ): a Boolean</h1>
+          1. Let _o_ be the *this* value.
+          1. Perform ? PrivateSet(_privateStateName_, _o_, _value_).
+          1. Return *undefined*.
+        1. Let _setter_ be ! CreateBuiltinFunction(_setterClosure_, 1, *""*, &laquo; &raquo;).
+        1. Perform MakeMethod(_setter_, _homeObject_).
+        1. Perform SetFunctionName(_setter_, _name_, *"set"*).
```
```suggestion
        1. Let _setter_ be ! CreateBuiltinFunction(_setterClosure_, 1, *"set"*, &laquo; &raquo;).
        1. Perform MakeMethod(_setter_, _homeObject_).
```

pzuraq commented:
@ljharb @jmdyck I believe I've addressed all the open issues so far, including the SDO and chain production changes. Let me know if it all looks good!

ljharb commented on spec.html:
```diff
@@ -24178,16 +24392,359 @@ <h1>Runtime Semantics: Evaluation</h1>
+    <emu-clause id="sec-createdecoratoraccessobject" type="abstract operation">
+      <h1>
+        CreateDecoratorAccessObject (
+          _kind_: *"field"*, *"accessor"*, *"method"*, *"getter"*, or *"setter"*,
```
Why are the strings being used here instead of the spec enums?

pzuraq replied:
The kinds pass through a couple of different AOs, so each AO would have to implement getting the string from the kind which seemed a bit repetitive. I suppose we could also add an AO for getting the string from the kind though, I'll do that.

pzuraq replied:
Actually I see now that we only really use the strings in one place, so will just inline the conversion there.

ljharb commented on spec.html:
```diff
@@ -24635,6 +25211,102 @@ <h1>Static Semantics: ContainsArguments ( ): a Boolean</h1>
+          1. Return *undefined*.
+        1. Let _setter_ be CreateBuiltinFunction(_setterClosure_, 1, *"set"*, &laquo; &raquo;).
+        1. Perform MakeMethod(_setter_, _homeObject_).
+        1. Perform SetFunctionName(_setter_, _name_, *"set"*).
```
```suggestion
```
this is already done in CreateBuiltinFunction

ljharb commented on spec.html:
```diff
@@ -24706,27 +25391,36 @@ <h1>
         Runtime Semantics: ClassElementEvaluation (
           _object_: unknown,
-        ): either a normal completion containing either a ClassFieldDefinition Record, a ClassStaticBlockDefinition Record, a Private Name, or ~unused~, or an abrupt completion
+          _metadataList: a List of Metadata Records,
```
```suggestion
          _metadataList_: a List of Metadata Records,
```
https://github.com/tc39/ecma262/pull/2417/files#r800188546

ljharb commented on spec.html:
```diff
@@ -24635,6 +25211,102 @@ <h1>Static Semantics: ContainsArguments ( ): a Boolean</h1>
+          1. Return ? PrivateGet(_privateStateName_, _o_).
+        1. Let _getter_ be CreateBuiltinFunction(_getterClosure_, 0, *"get"*, &laquo; &raquo;).
+        1. Perform MakeMethod(_getter_, _homeObject_).
+        1. Perform SetFunctionName(_getter_, _name_, *"get"*).
```
```suggestion
```

arhadthedev commented on spec.html:
```diff
@@ -1273,6 +1273,17 @@ <h1>Well-Known Symbols</h1>
                 An object valued property whose own and inherited property names are property names that are excluded from the `with` environment bindings of the associated object.
               </td>
             </tr>
+            <tr>
+              <td>
+                <dfn>@@metadata</dfn>
+              </td>
+              <td>
+                *"Symbol.metadata"*
+              </td>
+              <td>
+                An object valued property that contains the metadata associated with an object. Primarily used to store metadata defined by decorators on classes.
+              </td>
+            </tr>
```
I believe the table is sorted alphabetically so `@@metadata` should be between `@@matchAll` and `@@replace`.

arhadthedev commented on spec.html:
```diff
@@ -4652,16 +4663,19 @@ <h1>The PrivateElement Specification Type</h1>
-      <emu-table id="table-classfielddefinition-fields" caption="ClassFieldDefinition Record Fields">
+    <emu-clause id="sec-classelementdefinition-record-specification-type" oldids="sec-classfielddefinition-record-specification-type">
+      <h1>The ClassElementDefinition Record Specification Type</h1>
+      <p>The ClassElementDefinition type is a Record used in the specification of class element definitions.</p>
```
The definition repeats itself (The FooBarBaz [...] is a foo bar baz). Maybe it should be detalized as in `The PrivateElement type is a Record used in the specification of private class fields, methods, and accessors.`

arhadthedev commented on spec.html:
```diff
@@ -4682,13 +4699,165 @@ <h1>The ClassFieldDefinition Record Specification Type</h1>
           <tr>
             <td>
-              [[Initializer]]
+              [[Kind]]
```
Since the table is heavily expanded from two items to eight, here is a chance to order it alphabetically as well.

arhadthedev commented on spec.html:
```diff
@@ -4682,13 +4699,165 @@ <h1>The ClassFieldDefinition Record Specification Type</h1>
+      <p>The Metadata type is a Record used in the specification of decorator metadata. When defining metadata for a given decorated value, one Metadata Record is created for each user provided unique metadata key.</p>
+      <p>Values of the Metadata type are Record values whose fields are defined as by <emu-xref href="#table-metadata-fields"></emu-xref>. Such values are referred to as <dfn>Metadata Record</dfn>s.</p>
+      <emu-table id="table-metadata-fields" caption="Metadata Record Fields">
+        <table>
```
Ditto.

pzuraq commented:
Latest round of feedback addressed!

jmdyck commented on spec.html:
```diff
@@ -24649,17 +25320,30 @@ <h1>
+      <emu-alg>
+        1. Let _name_ be the result of evaluating |ClassElementName|.
+        1. ReturnIfAbrupt(_name_).
+        1. Let _privateStateDesc_ be the string-concatenation of *"accessor storage"*.
```
Is it supposed to be `the string-concatenation of _name_ and *" accessor storage"*`?

jmdyck commented on spec.html:
```diff
@@ -24819,55 +25515,95 @@ <h1>
-            1. If IsStatic of _e_ is *false*, append _element_ to _instanceFields_.
+            1. Return ? _result_.
+          1. Let _element_ be _result_.[[Value]].
+          1. Else if _element_ is a ClassElementDefinition Record, then
```
This is an `Else` without a preceding `If`!

jmdyck commented:
If you rebase to main, I can give you another PR of changes.

pzuraq commented:
Updates:

- Changed `initialize` to `init`, per https://github.com/tc39/proposal-decorators/issues/446
- Added state for tracking decorator application status, and added an error when calling context methods after decorator has been applied, per https://github.com/tc39/proposal-decorators/issues/447
- Addressed outstanding feedback

@jmdyck rebased, would love another round of fixes/updates 😄

ljharb commented on spec.html:
```diff
@@ -13193,24 +13416,24 @@ <h1>ECMAScript Function Objects</h1>
           </td>
           <td>
-            If the function is a class, this is a list representing the non-static private methods and accessors of the class.
+            If the function is a class with decorated prototype methods or accessors, this is a list of initializer functions added by the decorators, or ~empty~.
```
could this also apply to a class with a class-level decoration, but no decorated prototype methods or accessors? or, a class with only field decorators?

ljharb commented on spec.html:
```diff
@@ -24176,16 +24390,392 @@ <h1>Runtime Semantics: Evaluation</h1>
+      </dl>
+      <emu-alg>
+        1. Let _contextObj_ be OrdinaryObjectCreate(%Object.prototype%).
+        1. If _kind_ is ~method~, let _kindStr_ *"method"*.
+        1. Else if _kind_ is ~getter~, let _kindStr_ be *"getter"*.
+        1. Else if _kind_ is ~setter~, let _kindStr_ be *"setter"*.
+        1. Else if _kind_ is ~accessor~, let _kindStr_ be *"accessor"*.
+        1. Else if _kind_ is ~field~, let _kindStr_ be *"field"*.
+        1. Else,
+          1. Assert: _kind_ is ~class~.
+          1. Let _kindStr_ be *"class"*.
```
even if this logic is only used in one place, i lightly think it might deserve its own AO?

Jack-Works commented on spec.html:
```diff
@@ -24176,16 +24390,392 @@ <h1>Runtime Semantics: Evaluation</h1>
+
+      DecoratorCallExpression[Yield, Await] :
+        DecoratorMemberExpression[?Yield, ?Await] Arguments[?Yield, ?Await]
+        DecoratorParenthesizedExpression[?Yield, ?Await] Arguments[?Yield, ?Await]
```
```js
@(a[1])(args)
```

Why do we need this syntax rule? Why not just

```js
@(a[1](args))
```

pzuraq replied:
I’m matching the syntax which shipped in Babel here for stage 2 decorators, which is the syntax that was agreed upon before I started working on decorators. Personally wouldn’t mind changing it, I will discuss with the larger group.

Jack-Works commented on spec.html:
```diff
@@ -24176,16 +24390,392 @@ <h1>Runtime Semantics: Evaluation</h1>
+        1. Let _accessObj_ be OrdinaryObjectCreate(%Object.prototype%).
+        1. If _kind_ is ~field~, ~accessor~, ~method~, or ~getter~, then
+          1. Let _getterClosure_ be a new Abstract Closure with no parameters that captures _name_ and performs the following steps when called:
+            1. Let _o_ be the *this* value.
```
**this** will be the `accessObj`. Shouldn't it always bond to the class instance?

Jack-Works replied:
when it called as `accessObj.get()`

pzuraq replied:
It should not, the `accessObj` is created once per class, since decoration only occurs once per class. The user must call the function with `this` being the instance when actually using the function. This is an unintuitive but advanced API, and decorator authors are used to doing this already since decoration only occurs once on the prototype with all iterations of the proposal.

ljharb replied:
Getters, like all prototype methods, should be borrowable - both .call-able on other objects, and also should work with inheritance, all of which requires dynamically looking up the receiver.

Jack-Works replied:
so the correct usage of this API should be `accessObj.get.call(this)`? This is not intuitive...

pzuraq replied:
Correct, this is a technique which decorator authors use often. For instance, when decorating a method, you must call the original method this way:

```js
addOne(original) {
  return function() {
    return original.call(this) + 1;
  }
}
```

I agree that it is not intuitive, but metaprogramming rarely is in my experience. Importantly, this is only complicated for decorator _authors_. Decorator users, who are far more common in general, never have to work with these unintuitive APIs.

Jack-Works replied:
it can be more intuitive, like

```js
accessObj.get(classInstance)
```

and `accessObj.get()` default binds to the current instance

ljharb replied:
that would not be intuitive, because it's not how anything in the language behaves.

Jack-Works replied:
hmm, I don't expect writing a decorator needs to know `receivers`/`this` so clearly to know they should write `access.get.call(instance)`.

I think getting `this` from the first parameter, in this case, isn't too bad.

bathos replied:
There is also a purely syntactic way to perform this delegation when needed:

```js
function addOne(method) {
  return {
    __proto__: { method },
    method() {
      return super.method() + 1;
    }
  }.method;
}
```

(If `::` or similar proposals succeed, they will cover this scenario without the awkward [[HomeObject]] boilerplate.)

jmdyck commented:
> @jmdyck rebased, would love another round of fixes/updates smile

Okay, PR at https://github.com/pzuraq/ecma262/pull/3

jmdyck commented on spec.html:
```diff
@@ -24817,55 +25545,89 @@ <h1>
+        1. Set _F_.[[Elements]] to _instanceElements_.
+        1. Set _F_.[[Initializers]] to _instanceExtraInitializers_.
         1. For each PrivateElement _method_ of _staticPrivateMethods_, do
           1. Perform ! PrivateMethodOrAccessorAdd(_F_, _method_).
```
You removed the definition of `PrivateMethodOrAccessorAdd`, so shouldn't be calling it.

rbuckton replied:
It doesn't seem like the spec initializes static methods anymore. Was `InitializePrivateMethods` intended to replace this as well? If so, you're missing a call to it.

woess replied:
Apart from `PrivateMethodOrAccessorAdd` being missing, should static private methods be defined on the value of `F` before or after `ApplyDecoratorsToClassDefinition`? The current spec text seems to imply old `F`, but I think new `F` is the intended behavior and would make more sense.

pzuraq replied:
Hmm, that is tricky, the idea was that class decorators would run after method definition but before field definition. I think private methods work more like readonly fields though in a lot of ways, and it definitely doesn't make sense to have them be defined on a different object than fields, so I think you're correct, they should be defined after decorators have run.

woess replied:
Yeah, it's a tricky area and I haven't thought it through, but it seems clear to me that you want to define private static methods on the new `F`, which is also the _classBinding_, otherwise you can't access the method in the usual way (`C.#staticMethod()`).

jmdyck commented on spec.html:
```diff
@@ -13193,24 +13340,24 @@ <h1>ECMAScript Function Objects</h1>
           </td>
           <td>
-            List of PrivateElements
+            List of function objects | ~empty~
```
When would it be `~empty~`?

pzuraq replied:
It's set to `~empty~` after the decorators have been applied, mostly for bookkeeping purposes. It doesn't _really_ matter so I guess we could remove that, doesn't change any behavior.

jmdyck replied:
Hm. At the end of `ApplyDecoratorsToElementDefinition`, there's
`Set _elementRecord_.[[Decorators]] to ~empty~`,
but that's the `[[Decorators]]` field of a ClassElementDefinition Record, whereas I'm asking about the `[[Initializers]]` internal slot of an ECMAScript function object. I don't see the latter ever being set to `~empty~`.

jmdyck commented on spec.html:
```diff
@@ -24176,16 +24314,265 @@ <h1>Runtime Semantics: Evaluation</h1>
+      <h1>
+        CreateDecoratorAccessObject (
+          _kind_: ~field~, ~method~, ~accessor~, ~getter~, or ~setter~,
+          _name_: a String or Private Name,
```
```suggestion
          _name_: a String or a Private Name,
```

jmdyck commented on spec.html:
```diff
@@ -24176,16 +24314,265 @@ <h1>Runtime Semantics: Evaluation</h1>
+      <h1>
+        CreateDecoratorContextObject (
+          _kind_: ~class~, ~field~, ~method~, ~accessor~, ~getter~, or ~setter~,
+          _name_: a String, a Symbol, or Private Name,
```
```suggestion
          _name_: a String, a Symbol, or a Private Name,
```

jmdyck commented on spec.html:
```diff
@@ -24633,11 +25039,104 @@ <h1>Static Semantics: ContainsArguments ( ): a Boolean</h1>
+      <h1>
+        ApplyDecoratorsAndDefineMethod (
+          _homeObject_: an Object,
+          _methodDefinition_: a ClassElementDefinition record,
```
```suggestion
          _methodDefinition_: a ClassElementDefinition Record,
```

jmdyck commented on spec.html:
```diff
@@ -24633,11 +25039,104 @@ <h1>Static Semantics: ContainsArguments ( ): a Boolean</h1>
+        ApplyDecoratorsAndDefineMethod (
+          _homeObject_: an Object,
+          _methodDefinition_: a ClassElementDefinition record,
+          _extraInitializers_: A List of function objects,
```
```suggestion
          _extraInitializers_: a List of function objects,
```

pzuraq commented on spec.html:
```diff
@@ -24176,16 +24314,265 @@ <h1>Runtime Semantics: Evaluation</h1>
+        1. Else,
+          1. Perform ! CreateDataPropertyOrThrow(_contextObj_, *"name"*, _name_).
+        1. Let _addInitializer_ be CreateAddInitializerFunction(_initializers_, _decorationState_).
+        1. Perform ! CreateDataPropertyOrThrow(_contextObj_, *"addInitializer"*, _addInitializer_).
```
Yes this was a mistake, there is an open issue about it here: https://github.com/tc39/proposal-decorators/issues/464

So far the consensus seems to be that the spec does not need to change and the explainer can be updated. I think this may make the most sense.

ExE-Boss commented on spec.html:
```diff
@@ -24176,16 +24314,265 @@ <h1>Runtime Semantics: Evaluation</h1>
+            1. Else if _newValue_ is not *undefined*, throw a *TypeError* exception.
+          1. Else,
+            1. If IsCallable(_newValue_) is *true*, then
+              1. Perform MakeMethod(_newValue_, _homeObject_).
```
Note that this call will fail if <var>newValue</var> doesn’t have a [[HomeObject]] internal slot, which would be the case when <var>newValue</var> is a built‑in function, which is now also the case for a class that’s using a default constructor[^1].

[^1]: <https://github.com/tc39/ecma262/pull/2216>

---

In fact, the [[HomeObject]] should probably not be changed at all if it’s not `undefined`, otherwise it will cause weirdness, especially when <var>newValue</var> is a method taken from an existing class.

pzuraq replied:
This behavior is being discussed here as well: https://github.com/tc39/proposal-decorators/issues/497

I have it on the agenda for the next plenary to remove this line, as it does result in very odd behavior that does not seem desirable.

ljharb commented on spec.html:
```diff
@@ -24176,16 +24314,265 @@ <h1>Runtime Semantics: Evaluation</h1>
+          1. Else,
+            1. If IsCallable(_newValue_) is *true*, then
+              1. Perform MakeMethod(_newValue_, _homeObject_).
+              1. If _kind_ is ~getter~, then
+                1. Perform SetFunctionName(_newValue_, _key_, *"get"*).
+                1. Set _elementRecord_.[[Get]] to _newValue_.
+              1. Else if _kind_ is ~setter~, then
+                1. Perform SetFunctionName(_newValue_, _key_, *"set"*).
+                1. Set _elementRecord_.[[Set]] to _newValue_.
+              1. Else,
+                1. Perform SetFunctionName(_newValue_, _key_).
+                1. Set _elementRecord_.[[Value]] to _newValue_.
```
Ive always thought that assertion is kind of silly; couldn’t we just remove it?

JLHwung commented on spec.html:
```diff
@@ -24176,16 +24314,265 @@ <h1>Runtime Semantics: Evaluation</h1>
+
+      DecoratorMemberExpression[Yield, Await] :
+        IdentifierReference[?Yield, ?Await]
+        PrivateIdentifier
+        DecoratorMemberExpression[?Yield, ?Await] `.` IdentifierName
```
```suggestion
        DecoratorMemberExpression[?Yield, ?Await] `.` IdentifierName
        DecoratorMemberExpression[?Yield, ?Await] `.` PrivateIdentifier
```

Otherwise we would allow invalid `@#dec` and disallow valid `@C.#dec`.

pzuraq replied:
Yes that's correct I think, and given `@#dec` does not have any logical meaning I don't think this would be a normative change, so going to update the spec with the fix.

rbuckton commented on spec.html:
```diff
@@ -24176,16 +24314,265 @@ <h1>Runtime Semantics: Evaluation</h1>
+      <emu-alg>
+        1. For each element _decorator_ of _decorators_, do
+          1. Let _decorationState_ be the Record { [[Finished]]: *false* }.
+          1. Let _context_ be CreateDecoratorContextObject(~class~, _className_, _classExtraInitializers_, _decorationState_).
```
The _className_ variable isn't defined anywhere in this AO.

rbuckton replied:
Neither is _classExtraInitializers_. Should that be a reference to the _extraInitializers_ parameter?

acutmore commented on spec.html:
```diff
@@ -24176,16 +24314,265 @@ <h1>Runtime Semantics: Evaluation</h1>
+          1. Let _newValue_ be ? Call(_decorator_, *undefined*, &laquo; _value_, _context_ &raquo;).
+          1. Set _decorationState_.[[Finished]] to *true*.
+          1. If _kind_ is ~field~, then
+            1. If IsCallable(_initializer_) is *true*, append _initializer_ to _elementRecord_.[[Initializers]].
```
`_initializer_` being accessed here but is not defined.

woess commented on spec.html:
```diff
@@ -24817,55 +25337,79 @@ <h1>
+        1. Let _instanceExtraInitializers_ be a new empty List.
+        1. Let _staticExtraInitializers_ be a new empty List.
+        1. For each element _e_ of _staticElements_, do
+          1. If _e_.[[Kind]] is not ~field~, then
```
This assumes that _e_ is a ClassElementDefinition Record; however, it could also be a ClassStaticBlockDefinition Record (see a few lines above). So this should probably say sth like:
> If _e_ is a ClassElementDefinition Record and _e_.[[Kind]] is not field, then

pzuraq commented on spec.html:
```diff
@@ -24176,16 +24314,265 @@ <h1>Runtime Semantics: Evaluation</h1>
+      <emu-alg>
+        1. Let _addInitializerClosure_ be a new Abstract Closure with parameters (_initializer_) that captures _initializers_ and _decorationState_ and performs the following steps when called:
+          1. If _decorationState_.[[Finished]] is *true*, throw a *TypeError* exception.
+          1. Append _initializer_ to _initializers_.
```
Is this the type of change that requires consensus? Technically it would result in different execution behavior, but I think that it would be correct to do so and don't think passing another type of value would be logically coherent

ljharb replied:
Yes, that would be normative, but i doubt it would be controversial.

pzuraq commented on spec.html:
```diff
@@ -24176,16 +24314,265 @@ <h1>Runtime Semantics: Evaluation</h1>
+          1. Let _setterClosure_ be a new Abstract Closure with parameters (_value_) that captures _name_ and performs the following steps when called:
+            1. Let _o_ be the *this* value.
+            1. If _name_ is a String, perform ? PrivateSet(_name_, _o_, _value_).
+            1. Else, perform ? Set(_o_, _name_, _value_, *true*).
```
This has been addressed in the latest changes

pzuraq commented:
PR has been rebased, I squashed the existing commits to make rebasing easier overall. The only new commit contains non-normative fixes. The remaining issues open in this thread are all normative I believe, and I will be opening up separate PRs for each of them so we can discuss at the upcoming plenary.

jmdyck commented on spec.html:
```diff
@@ -6693,21 +6786,55 @@ <h1>
+        1. Let _privateMethods_ be a new empty List.
+        1. For each element _element_ of _elementDefinitions_, do
+          1. If _element_.[[Key]] is a Private Name and _element_.[[Kind]] is ~method~, ~getter~, ~setter~, or ~accessor~, then
+            1. If _privateMethods_ contains a PrivateElement whose [[Key]] is _element_.[[Key]], then
+              1. Let _existing_ be that PrivateElement.
```
```suggestion
            1. If _privateMethods_ contains a PrivateElement _existing_ such that _existing.[[Key]] is _element_.[[Key]], then
```

jmdyck commented on spec.html:
```diff
@@ -23991,16 +24145,273 @@ <h1>Runtime Semantics: Evaluation</h1>
+      <emu-grammar>Decorator : `@` DecoratorMemberExpression</emu-grammar>
+      <emu-alg>
+        1. Let _expr_ be the |MemberExpression| that is covered by |DecoratorMemberExpression|.
+        1. Let _ref_ be the result of evaluating _expr_.
```
```suggestion
        1. Let _ref_ be ? Evaluation of _expr_.
```

jmdyck commented on spec.html:
```diff
@@ -23991,16 +24145,273 @@ <h1>Runtime Semantics: Evaluation</h1>
+      <emu-grammar>Decorator : `@` DecoratorCallExpression</emu-grammar>
+      <emu-alg>
+        1. Let _expr_ be the |CallMemberExpression| that is covered by |DecoratorCallExpression|.
+        1. Let _ref_ be the result of evaluating _expr_.
```
```suggestion
        1. Let _ref_ be ? Evaluation of _expr_.
```

jmdyck commented on spec.html:
```diff
@@ -23991,16 +24145,273 @@ <h1>Runtime Semantics: Evaluation</h1>
+      <emu-grammar>Decorator : `@` DecoratorParenthesizedExpression</emu-grammar>
+      <emu-alg>
+        1. Let _expr_ be the |MemberExpression| that is covered by |DecoratorParenthesizedExpression|.
+        1. Let _ref_ be the result of evaluating _expr_.
```
```suggestion
        1. Let _ref_ be ? Evaluation of _expr_.
```

jmdyck commented on spec.html:
```diff
@@ -24461,17 +24984,30 @@ <h1>
+        1. Let _name_ be the result of evaluating |ClassElementName|.
+        1. ReturnIfAbrupt(_name_).
+        1. Let _privateStateDesc_ be the string-concatenation of _name_ and *" accessor storage"*.
+        1. Let _privateStateName_ be a new Private Name whose [[Description]] value is _privateStateDesc_.
```
```suggestion
        1. Let _privateStateName_ be a new Private Name whose [[Description]] is _privateStateDesc_.
```

jmdyck commented on spec.html:
```diff
@@ -28107,6 +28688,21 @@ <h1>Static Semantics: Early Errors</h1>
+            It is a Syntax Error if |DecoratorList| is present and |Declaration| is not |ClassDeclaration|.
+          </li>
+          <li>
+            It is a Syntax Error if |DecoratorList| is present and |Declaration| is a |ClassDeclaration| and the |DecoratorList| of that |ClassDeclaration| is present.
```
```suggestion
            It is a Syntax Error if |DecoratorList| is present, |Declaration| is a |ClassDeclaration|, and the |DecoratorList| of that |ClassDeclaration| is present.
```

jmdyck commented on spec.html:
```diff
@@ -47180,6 +47785,11 @@ <h1>Functions and Classes</h1>
+    <emu-prodref name="DecoratorList"></emu-prodref>
+    <emu-prodref name="Decorator"></emu-prodref>
+    <emu-prodref name="DecoratorMemberExpression"></emu-prodref>
+    <emu-prodref name="DecoratorCallExpression"></emu-prodref>
+    <emu-prodref name="DecoratorParenthesizedExpression"></emu-prodref>
```
Swap the order of these two lines.

pzuraq commented on spec.html:
```diff
@@ -24448,11 +24878,104 @@ <h1>Static Semantics: ContainsArguments ( ): a Boolean</h1>
+      <h1>
+        MakeAutoAccessorGetter (
+          _homeObject_: an Object,
+          _name_: a property key or Private Name,
```
Should have been used in `SetFunctionName`, added

pzuraq commented on spec.html:
```diff
@@ -6693,21 +6786,55 @@ <h1>
       </emu-alg>
     </emu-clause>
 
-    <emu-clause id="sec-definefield" type="abstract operation">
```
Added

ljharb commented on spec.html:
```diff
@@ -13403,21 +13564,23 @@ <h1>
+        1. Assert: _methodDefinition_.[[Kind]] is ~method~, ~getter~, ~setter~, or ~accessor~.
+        1. Let _key_ be _methodDefinition_.[[Key]].
+        1. If _key_ is not a Private Name, then
+          1. Let _desc_ be the PropertyDescriptor { [[Writable]]: *true*, [[Enumerable]]: _enumerable_, [[Configurable]]: *true* }.
+          1. If _methodDefinition_.[[Kind]] is ~getter~ or ~accessor~, set _desc_.[[Get]] to _methodDefinition_.[[Get]].
+          1. If _methodDefinition_.[[Kind]] is ~setter~ or ~accessor~, set _desc_.[[Set]] to _methodDefinition_.[[Set]].
+          1. If _methodDefinition_.[[Kind]] is ~method~, set _desc_.[[Value]] to _methodDefinition_.[[Value]].
```
[[Writable]] is invalid on an accessor descriptor.
```suggestion
          1. Let _desc_ be the PropertyDescriptor { [[Enumerable]]: _enumerable_, [[Configurable]]: *true* }.
          1. If _methodDefinition_.[[Kind]] is ~getter~ or ~accessor~, set _desc_.[[Get]] to _methodDefinition_.[[Get]].
          1. If _methodDefinition_.[[Kind]] is ~setter~ or ~accessor~, set _desc_.[[Set]] to _methodDefinition_.[[Set]].
          1. If _methodDefinition_.[[Kind]] is ~method~, set _desc_.[[Value]] to _methodDefinition_.[[Value]] and _desc_.[[Writable]] to *true*.
```

ljharb commented on spec.html:
```diff
@@ -23991,16 +24144,273 @@ <h1>Runtime Semantics: Evaluation</h1>
+        1. If _kind_ is ~field~, ~method~, ~accessor~, or ~getter~, then
+          1. Let _getterClosure_ be a new Abstract Closure with parameters (_obj_) that captures _name_ and performs the following steps when called:
+            1. If _obj_ is not an Object, throw a *TypeError* exception.
+            1. If _name_ is either a String or a Symbol, return ? Get(_obj_, _name_).
```
```suggestion
            1. If _name_ is a property key, return ? Get(_obj_, _name_).
```

ljharb commented on spec.html:
```diff
@@ -23991,16 +24144,273 @@ <h1>Runtime Semantics: Evaluation</h1>
+        1. If _kind_ is ~field~, ~accessor~, or ~setter~, then
+          1. Let _setterClosure_ be a new Abstract Closure with parameters (_obj_, _value_) that captures _name_ and performs the following steps when called:
+            1. If _obj_ is not an Object, throw a *TypeError* exception.
+            1. If _name_ is either a String or a Symbol, perform ? Set(_obj_, _name_, _value_, *true*).
```
```suggestion
            1. If _name_ is a property key, perform ? Set(_obj_, _name_, _value_, *true*).
```

ljharb commented on spec.html:
```diff
@@ -23991,16 +24144,273 @@ <h1>Runtime Semantics: Evaluation</h1>
+          1. Perform ! CreateDataPropertyOrThrow(_accessObj_, *"set"*, _setter_).
+        1. Let _hasClosure_ be a new Abstract Closure with parameters (_obj_) that captures _name_ and performs the following steps when called:
+          1. If _obj_ is not an Object, throw a *TypeError* exception.
+          1. If _name_ is either a String or a Symbol, return ? HasProperty(_obj_, _name_).
```
```suggestion
          1. If _name_ is a property key, return ? HasProperty(_obj_, _name_).
```

ljharb commented on spec.html:
```diff
@@ -23991,16 +24144,273 @@ <h1>Runtime Semantics: Evaluation</h1>
+          1. If _name_ is either a String or a Symbol, return ? HasProperty(_obj_, _name_).
+          1. If PrivateElementFind(_obj_, _name_) is not ~empty~, return *true*.
+          1. Return *false*.
+        1. Let _has_ be CreateBuiltinFunction(_hasClosure_, 1, *""*, « »).
```
```suggestion
        1. Let _has_ be CreateBuiltinFunction(_hasClosure_, 1, *"has"*, « »).
```
perhaps?

ljharb commented on spec.html:
```diff
@@ -23991,16 +24144,273 @@ <h1>Runtime Semantics: Evaluation</h1>
+      <h1>
+        CreateDecoratorContextObject (
+          _kind_: ~class~, ~field~, ~method~, ~accessor~, ~getter~, or ~setter~,
+          _name_: a String, a Symbol, or a Private Name,
```
```suggestion
          _name_: a property key or a Private Name,
```

ljharb commented on spec.html:
```diff
@@ -23991,16 +24144,273 @@ <h1>Runtime Semantics: Evaluation</h1>
+          1. Let _getterClosure_ be a new Abstract Closure with parameters (_obj_) that captures _name_ and performs the following steps when called:
+            1. If _obj_ is not an Object, throw a *TypeError* exception.
+            1. If _name_ is either a String or a Symbol, return ? Get(_obj_, _name_).
+            1. Else, return ? PrivateGet(_name_, _obj_).
```
i just noticed this, but it seems really weird to have PrivateGet/Set take name first instead of object first (like Get/Set do).

jmdyck replied:
PrivateGet/Set *do* take the object first. This invocation has the args in the wrong order.

doehyunbaek commented:
Hi, esmeta typecheck is failing because `InitializeInstanceElements`, which calls `InitializePrivateMethods`, expects to return `throw completion`, as exmplified [here](https://tc39.es/ecma262/#sec-initializeinstanceelements).

`InitializePrivateMethods`, on the other hand, declares it might return `abrupt completion`, which is not a subtype of `throw completion`, as seen [here](https://github.com/tc39/ecma262/pull/2417/files#diff-181371b08d71216599b0acccbaabd03c306da6de142ea6275c2135810999805aR6794).

Changing the return type of `InitializePrivateMethods`, as this commit https://github.com/doehyunbaek/ecma262/commit/8a2b51d858aed89a45503d99b780d3d39c538ba2 actually solves the issue.

jmdyck commented on spec.html:
```diff
@@ -6693,21 +6786,54 @@ <h1>
-          1. Let _initValue_ be ? Call(_initializer_, _receiver_).
-        1. Else, let _initValue_ be *undefined*.
+        1. Assert: _elementRecord_.[[Kind]] is ~field~ or ~accessor~.
+        1. If _elementRecord_.[[BackingStorageKey]] is present, let _fieldName_ be _elementRecord_.[[BackingStorageKey]].
```
```suggestion
        1. If _elementRecord_ has a [[BackingStorageKey]] field, let _fieldName_ be _elementRecord_.[[BackingStorageKey]].
```

jmdyck commented on spec.html:
```diff
@@ -9057,9 +9184,12 @@ <h1>
+        1. If |DecoratorList?| is present, then
+          1. Let _decorators_ be ? DecoratorListEvaluation of |DecoratorList|.
+        1. Else, let _decorators_ be a new empty List.
+        1. Let _value_ be ? ClassDefinitionEvaluation of |ClassTail| with arguments *undefined*, _name_, and _decorators_.
```
ClassDefinitionEvaluation requires its 3rd arg to be a List of function objects, but I don't think there's anything here to ensure that `_decorators_` is such a List. Similarly for some other invocations of ClassDefinitionEvaluation.

(Maybe ClassDefinitionEvalution should accept a List of ES language values, and then deal with the possibility that one or more might not be a function object.)

(Similarly, BindingClassDeclarationEvaluation should maybe accept a List of ES language values.)

pzuraq replied:
The list is actually of `DecoratorDefinition` records since decorators need to have their receiver and the value. I've updated the `DecoratorDefinition` record to store any ECMAScript language value, and when that value is passed to `Call` during application that will throw a type error if it is not callable.

jmdyck commented on spec.html:
```diff
@@ -23991,16 +24144,273 @@ <h1>Runtime Semantics: Evaluation</h1>
+          1. Let _setterClosure_ be a new Abstract Closure with parameters (_obj_, _value_) that captures _name_ and performs the following steps when called:
+            1. If _obj_ is not an Object, throw a *TypeError* exception.
+            1. If _name_ is either a String or a Symbol, perform ? Set(_obj_, _name_, _value_, *true*).
+            1. Else, perform ? PrivateSet(_name_, _obj_, _value_).
```
```suggestion
            1. Else, perform ? PrivateSet(_obj_, _name_, _value_).
```

jmdyck commented on spec.html:
```diff
@@ -24448,11 +24877,106 @@ <h1>Static Semantics: ContainsArguments ( ): a Boolean</h1>
+      <emu-alg>
+        1. Let _getterClosure_ be a new Abstract Closure with no parameters that captures _privateStateName_ and performs the following steps when called:
+          1. Let _o_ be the *this* value.
+          1. Return ? PrivateGet(_privateStateName_, _o_).
```
```suggestion
          1. Return ? PrivateGet(_o_, _privateStateName_).
```

jmdyck replied:
Also, PrivateGet requires an Object, but I don't think it's guaranteed that `_o_` is an Object.

(And similarly for `_o_` in MakeAutoAccessorSetter.)

jmdyck commented on spec.html:
```diff
@@ -24448,11 +24877,106 @@ <h1>Static Semantics: ContainsArguments ( ): a Boolean</h1>
+      <emu-alg>
+        1. Let _setterClosure_ be a new Abstract Closure with parameters (_value_) that captures _privateStateName_ and performs the following steps when called:
+          1. Let _o_ be the *this* value.
+          1. Perform ? PrivateSet(_privateStateName_, _o_, _value_).
```
```suggestion
          1. Perform ? PrivateSet(_o_, _privateStateName_, _value_).
```

pgoforth commented on spec.html:
```diff
@@ -23991,16 +24185,277 @@ <h1>Runtime Semantics: Evaluation</h1>
+              1. If IsCallable(_newSetter_) is *true*, set _elementRecord_.[[Set]] to _newSetter_.
+              1. Else if _newSetter_ is not *undefined*, throw a *TypeError* exception.
+              1. Let _initializer_ be ? Get(_newValue_, *"init"*).
+              1. If IsCallable(_initializer_) is *true*, append _initializer_ to _elementRecord_.[[Initializers]].
```
I [opened an issue](https://github.com/babel/babel/issues/15608) against `@babel/plugin-proposal-decorators` thinking they had implemented this spec incorrectly. They in turned pointed me to this line saying that they were indeed implementing the spec appropriately. As you can see from the issue description, appending initializers results in an inside-out evaluation of the decorator's initializers, while everything else is evaluated outside-in (top->bottom, left-right).

Shouldn't initializers be prepended onto `[[Initializers]]` in order to maintain the same outside-in execution order?

pzuraq replied:
@pgoforth this was intentional. The ordering is from innermost decorator to outermost. This means innermost initializers run before outermost, and wrapped methods are wrapped from innermost to outermost, with the outermost method becoming the final method on the class. I commented on the other issue illustrating how the ordering of wrapped methods is actually not really predetermined - the outer method chooses when to call the inner. But the wrapping is consistent with the other orderings.

pgoforth replied:
@pzuraq Your example will give you the correct logging, but in a practical example the ordering is problematic and does not adhere to the expectations of the spec.

I have created a minimum repro with the issue in question [here](https://github.com/pgoforth/js-decorator-ordering-tests). To sum up, you have to create an overly complex decorator that inverts the order of the setter calls in order to maintain the original order of the initializer calls. I hope you can appreciate how confusing it is having a simple decorator giving an incorrect result because of innermost -> outermost execution.

ljharb replied:
Decorators are *evaluated* left to right top to bottom, but they are *applied* innermost to outermost, and that's certainly what I expect.

pgoforth replied:
@ljharb I think I'm not being clear. I have no problem with the ordering, that's irrelevant to my complaint. If the setters were applied innermost to outermost, why wouldn't the evalutation from left to right top to bottom provide the inner setter with the reference to the outer setter? Currently the outer setter can only call in the opposite direction that the initializers are applied in.

```js
function appendString(strToAppend) {
    return function ({ set }) {
        return {
            set(value) {
                // outtermost call is applied first first
                return set.call(this, `${value}${strToAppend}`);
            },
            init(initialValue) {
                // innermost call is applied first first
                return `${initialValue}${strToAppend}`
            },
        }
    }
}

class FakeClass{
    @appendString('Baz')
    @appendString('Bar')
    accessor fakeField = 'foo'
}
const instance = new FakeClassSimple()
console.log(instance.fakeField) // 'fooBarBaz'
// Setting to the same value it's initialized to, results in the opposite result
instance.fakeField = 'foo'
console.log(instance.fakeField) // 'fooBazBar'
```

Can you see how this is confusing?

pzuraq replied:
This convo has been resolved with a change in the init order in the spec

pzuraq committed:
Rebase and update based on feedback

pzuraq commented on spec.html:
```diff
@@ -24461,17 +24985,30 @@ <h1>
+      <emu-alg>
+        1. Let _name_ be the result of evaluating |ClassElementName|.
+        1. ReturnIfAbrupt(_name_).
+        1. Let _privateStateDesc_ be the string-concatenation of _name_ and *" accessor storage"*.
```
Updated to check the type and get the string value from the appropriate source

pzuraq commented on spec.html:
```diff
@@ -6693,21 +6786,54 @@ <h1>
+            1. If _privateMethods_ contains a PrivateElement _existing_ such that _existing.[[Key]] is _element_.[[Key]], then
+              1. Assert: _element_.[[Kind]] is ~getter~ or ~setter~.
+              1. Assert: _existing_.[[Kind]] is ~accessor~.
+              1. If _element_.[[Get]] is *undefined*, then
```
Refactored this to be clearer and to only use and define properties that should exist on the respective elements based on `[[Kind]]`

pzuraq commented:
PR has been updated and rebased, and all changes have been addressed!

jmdyck commented on spec.html:
```diff
@@ -13578,21 +13797,23 @@ <h1>
+        1. If _key_ is not a Private Name, then
+          1. Let _desc_ be the PropertyDescriptor { [[Enumerable]]: _enumerable_, [[Configurable]]: *true* }.
+          1. If _methodDefinition_.[[Kind]] is ~getter~ or ~accessor~, set _desc_.[[Get]] to _methodDefinition_.[[Get]].
+          1. If _methodDefinition_.[[Kind]] is ~setter~ or ~accessor~, set _desc_.[[Set]] to _methodDefinition_.[[Set]].
```
```suggestion
          1. If _methodDefinition_.[[Kind]] is either ~setter~ or ~accessor~, set _desc_.[[Set]] to _methodDefinition_.[[Set]].
```

jmdyck commented on spec.html:
```diff
@@ -13578,21 +13797,23 @@ <h1>
+          1. Let _desc_ be the PropertyDescriptor { [[Enumerable]]: _enumerable_, [[Configurable]]: *true* }.
+          1. If _methodDefinition_.[[Kind]] is ~getter~ or ~accessor~, set _desc_.[[Get]] to _methodDefinition_.[[Get]].
+          1. If _methodDefinition_.[[Kind]] is ~setter~ or ~accessor~, set _desc_.[[Set]] to _methodDefinition_.[[Set]].
+          1. If _methodDefinition_.[[Kind]] is ~method~, set _desc_.[[Value]] to _methodDefinition_.[[Value]] and _desc_.[[Writable]] to *true*.
```
The spec never uses the phrasing "set A to B and C to D".
```suggestion
          1. If _methodDefinition_.[[Kind]] is ~method~, then
            1. Set _desc_.[[Value]] to _methodDefinition_.[[Value]].
            1. Set _desc_.[[Writable]] to *true*.
```

jmdyck commented on spec.html:
```diff
@@ -24320,16 +24531,287 @@ <h1>Runtime Semantics: Evaluation</h1>
+              1. If IsCallable(_initializer_) is *true*, prepend _initializer_ to _elementRecord_.[[Initializers]].
+              1. Else if _initializer_ is not *undefined*, throw a *TypeError* exception.
+            1. Else if _newValue_ is not *undefined*, then
+              1. throw a *TypeError* exception.
```
```suggestion
              1. Throw a *TypeError* exception.
```

jmdyck commented on spec.html:
```diff
@@ -24320,16 +24531,287 @@ <h1>Runtime Semantics: Evaluation</h1>
+              1. Else,
+                1. Set _elementRecord_.[[Value]] to _newValue_.
+            1. Else if _newValue_ is not *undefined*, then
+              1. throw a *TypeError* exception.
```
```suggestion
              1. Throw a *TypeError* exception.
```

jmdyck commented on spec.html:
```diff
@@ -24790,17 +25386,34 @@ <h1>
+      <emu-alg>
+        1. Let _name_ be ? Evaluation of |ClassElementName|.
+        1. If _name_ is a Private Name, let _readableName_ be _name_.[[Description]].
+        1. Else if _name_ is a Symbol, let _readableName_ be SymbolDescriptiveString(_name).
```
```suggestion
        1. Else if _name_ is a Symbol, let _readableName_ be SymbolDescriptiveString(_name_).
```

pzuraq commented on spec.html:
```diff
@@ -6726,27 +6869,69 @@ <h1>
+              1. If _element_.[[Kind]] is ~setter~, let _setter_ be _element_.[[Set]].
+              1. Else, let _setter_ be *undefined*.
+              1. Let _existing_ be *undefined*.
+              1. If _privateMethods_ contains a PrivateElement _e_ such that _existing_.[[Key]] is _e_.[[Key]], then
```
Ah yes, good catch, that was the intent

pzuraq commented on spec.html:
```diff
@@ -4711,13 +4753,114 @@ <h1>The ClassFieldDefinition Record Specification Type</h1>
+              ~field~, ~method~, ~accessor~, ~getter~, ~setter~
+            </td>
+            <td>
+              ~field~, ~method~, ~accessor~, ~getter~, ~setter~
```
This is the opposite of the change that @ljharb requested [here](https://github.com/tc39/ecma262/pull/2417#discussion_r1128549769)

jmdyck replied:
My suggestion is based on PR #2877, or more specifically PR #2972.

pzuraq replied:
Ok, I added it back and updated it everywhere in that table per @ljharb's other suggestion

pzuraq committed:
Update re: feedback

pzuraq commented:
@jmdyck all issues resolved I believe, except the one that @ljharb disagreed on. Happy to change it either way, just want to make sure it's the right way.

pzuraq committed:
fix throw completion

pzuraq committed:
add back and/or

pzuraq commented on spec.html:
```diff
@@ -24623,10 +25134,103 @@ <h1>Static Semantics: ContainsArguments</h1>
+        1. Let _scope_ be the LexicalEnvironment of the running execution context.
+        1. Let _privateScope_ be the running execution context's PrivateEnvironment.
+        1. Let _sourceText_ be the empty sequence of Unicode code points.
+        1. Let _initializer_ be ! OrdinaryFunctionCreate(%Function.prototype%, _sourceText_, _formalParameterList_, _Initializer_, ~non-lexical-this~, _scope_, _privateScope_).
```
Is this still an issue? Not sure I understood it, I would think that this is a syntax error

pzuraq commented on spec.html:
```diff
@@ -24166,16 +24373,319 @@ <h1>Runtime Semantics: Evaluation</h1>
+      Decorator[Yield, Await] :
+        `@` DecoratorMemberExpression[?Yield, ?Await]
+        `@` DecoratorCallExpression[?Yield, ?Await]
+
```
I believe this should be fixed

pzuraq commented on spec.html:
```diff
@@ -24166,16 +24373,319 @@ <h1>Runtime Semantics: Evaluation</h1>
+        `(` Expression[+In, ?Yield, ?Await] `)`
+
+      DecoratorCallExpression[Yield, Await] :
+        DecoratorMemberExpression[?Yield, ?Await] Arguments[?Yield, ?Await]
```
I believe this should be fixed

ljharb commented on spec.html:
```diff
@@ -4681,16 +4681,55 @@ <h1>The PrivateElement Specification Type</h1>
+    <emu-clause id="sec-decoratordefinition-record-specification-type">
+      <h1>The DecoratorDefinition Record Specification Type</h1>
+      <p>
+        The DecoratorDefinition type is a Record used in the specification of
+        decorators.
```
```suggestion
        The DecoratorDefinition type is a Record used in the specification of decorators.
```

ljharb commented on spec.html:
```diff
@@ -4681,16 +4681,55 @@ <h1>The PrivateElement Specification Type</h1>
+      </p>
+      <p>
+        Values of the DecoratorDefinition type are Record values whose fields
+        are defined by
+        <emu-xref href="#table-decoratordefinition-fields"></emu-xref>. Such
+        values are referred to as
+        <dfn variants="DecoratorDefinition Record">DecoratorDefinition Records</dfn>.
```
i think we soft-wrap pretty consistently, not hard-wrap

ljharb commented on spec.html:
```diff
@@ -24320,16 +24533,287 @@ <h1>Runtime Semantics: Evaluation</h1>
+      </dl>
+      <emu-alg>
+        1. Let _contextObj_ be OrdinaryObjectCreate(%Object.prototype%).
+        1. If _kind_ is ~method~, then
+          1. Let _kindStr_ be *"method"*.
+        1. Else if _kind_ is ~getter~, then
+          1. Let _kindStr_ be *"getter"*.
+        1. Else if _kind_ is ~setter~, then
+          1. Let _kindStr_ be *"setter"*.
+        1. Else if _kind_ is ~accessor~, then
+          1. Let _kindStr_ be *"accessor"*.
+        1. Else if _kind_ is ~field~, then
+          1. Let _kindStr_ be *"field"*.
+        1. Else,
+          1. Assert: _kind_ is ~class~.
+          1. Let _kindStr_ be *"class"*.
```
aside for the editors: kind of feels like we need a `SpecValueToString()` AO or something for cases like this. i think atomics has one too.

lpardosixtosMs commented on spec.html:
```diff
@@ -6726,27 +6869,69 @@ <h1>
+              1. Else, let _getter_ be *undefined*.
+              1. If _element_.[[Kind]] is ~setter~, let _setter_ be _element_.[[Set]].
+              1. Else, let _setter_ be *undefined*.
+              1. Let _existing_ be *undefined*.
+              1. If _privateMethods_ contains a PrivateElement _e_ such that _e_.[[Key]] is _element_.[[Key]], then
+                1. Assert: _e_.[[Kind]] is ~accessor~.
+                1. Set _existing_ to _e_.
+                1. If _e_.[[Get]] is not *undefined*, set _getter_ to _existing_.[[Get]].
+                1. If _e_.[[Set]] is not *undefined*, set _setter_ to _existing_.[[Set]].
+              1. Let _privateElement_ be PrivateElement { [[Key]]: _element_.[[Key]], [[Kind]]: ~accessor~, [[Get]]: _getter_, [[Set]]: _setter_ }.
+              1. If _existing_ is not *undefined*, replace _existing_ in _privateMethods_ with _privateElement_.
```
I understand that steps 6 through 9 define a no-op if there is already a private accessor pair with the same key, is this correct? If so, could we reword 7.c and 7.d? It seems a little strange to use `_e_` and `_existing_` in the same line when `_existing_` was set to `_e_` in the previous step.

lpardosixtosMs commented on spec.html:
```diff
@@ -24320,16 +24533,287 @@ <h1>Runtime Semantics: Evaluation</h1>
+    <emu-clause id="sec-applydecoratorstoelementdefinition" type="abstract operation">
+      <h1>
+        ApplyDecoratorsToElementDefinition (
+          _homeObject_: an Object,
```
Looks like `_homeObject_` is not used here.
